public/app/features/live/live.ts

```diff
@@ -141,20 +161,163 @@ export class CentrifugeSrv implements GrafanaLiveSrv {
+        key = `xstr/${streamCounter++}`;
+      }
+
+      const process = (msg: DataFrameJSON) => {
+        if (!data) {
+          data = new StreamingDataFrame(msg, options.buffer);
+        } else {
+          data.push(msg);
+        }
+        state = LoadingState.Streaming;
+
+        // Filter out fields
+        if (!filtered || msg.schema) {
+          filtered = data;
+          if (options.filter) {
+            const { fields } = options.filter;
+            if (fields?.length) {
+              filtered = {
+                ...data,
+                fields: data.fields.filter((f) => fields.includes(f.name)),
+              };
+            }
+          }
+        }
+
+        const elapsed = perf.last - last;
+        if (elapsed > 1000 || perf.ok) {
+          filtered.length = data.length; // make sure they stay up-to-date
+          subscriber.next({ state, data: [filtered], key });
+          last = perf.last;
+        }
+      };
+
+      const sub = channel.getStream().subscribe({
+        error: (err: any) => {
+          console.log('LiveQuery [error]', { err }, options.addr);
+          state = LoadingState.Error;
+          subscriber.next({ state, data: [data], key, error: toDataQueryError(err) });
+          sub.unsubscribe(); // close after error
+        },
+        complete: () => {
+          console.log('LiveQuery [complete]', options.addr);
+          if (state !== LoadingState.Error) {
+            state = LoadingState.Done;
+          }
+          // or track errors? subscriber.next({ state, data: [data], key });
+          subscriber.complete();
+          sub.unsubscribe();
+        },
+        next: (evt: LiveChannelEvent) => {
+          if (isLiveChannelMessageEvent(evt)) {
+            process(evt.message);
+            return;
+          }
+          if (isLiveChannelStatusEvent(evt)) {
+            if (evt.error) {
+              let error = toDataQueryError(evt.error);
+              error.message = `Streaming channel error: ${error.message}`;
+              state = LoadingState.Error;
+              subscriber.next({ state, data: [data], key, error });
+              return;
+            } else if (
+              evt.state === LiveChannelConnectionState.Connected ||
+              evt.state === LiveChannelConnectionState.Pending
+            ) {
+              if (evt.message) {
+                process(evt.message);
+              }
+              return;
+            }
+            console.log('ignore state', evt);
+          }
+        },
+      });
+
+      return () => {
+        sub.unsubscribe();
+      };
+    });
+  }
+
+  /**
+   * For channels that support presence, this will request the current state from the server.
+   *
+   * Join and leave messages will be sent to the open stream
+   */
+  getPresence(address: LiveChannelAddress): Promise<LiveChannelPresenceStatus> {
+    return this.getChannel(address).getPresence();
   }
 }
 
+// incremet the stream ids
+let streamCounter = 10;
```

Comment:
Just curious - why 10 here?

Reply:
could be anything -- I'll change to zero